File: LeetCode/Math/Rust/Leetcode1175-prime-arrangements.rs
impl Solution {
    pub fn num_prime_arrangements(n: i32) -> i32 {
        let (P, mut st, mut primes) = (1e9 as i64 + 7, vec![false; n as usize + 1], vec![]);
        for i in 2..=n {
            if !st[i as usize] {
                primes.push(i);
            }
            let mut j = 0;
            while primes[j] <= n / i {
                st[(primes[j] * i) as usize] = true;
                if i % primes[j] == 0 {
                    break;
                }
                j += 1;
            }
        }
        let cnt = primes.len();
        let fact = |n: i32| -> i64 {
            let mut res: i64 = 1;
            for i in 1..=n {
                res = res * i as i64 % P;
            }
            res
        };

        (fact(n - cnt as i32) * fact(cnt as i32) % P) as i32
    }
}

File: LeetCode/Math/Python/Leetcode1175-prime-arrangements.py
class Solution:
    def numPrimeArrangements(self, n: int) -> int:
        st, primes, MOD = [False] * (n + 1), [], int(1e9 + 7)

        def fact(n):
            res = 1
            for i in range(1, n + 1):
                res = res * i % MOD
            return res

        for i in range(2, n + 1):
            if not st[i]: primes.append(i)
            j = 0
            while primes[j] * i <= n:
                st[primes[j] * i] = True
                if i % primes[j] == 0: break
                j += 1
        cnt = len(primes)
        # print(fact(n - cnt))
        return fact(n - cnt) * fact(cnt) % MOD
